### ISSUE TITLE
MuSiC2 issue: Error in MOD0 + mod0 : non-conformable arrays

### ISSUE THREAD

**Issue by jaclynbeck-sage** [xuranw/MuSiC#106]
Running `music2_prop`, the following issue happens if 'select.ct' argument = NULL. It's allowed to be NULL in the original music_prop, and a NULL value means "use all cell types in the metadata" for that function. The documentation for `music2_prop` also says this. However there's no check or logic for the case where select.ct = NULL in `music2_prop`, so when it gets to the iteration loop it creates a MOD0 matrix with 0 rows:

ncell = length(select.ct) # returns 0 for select.ct = NULL
...
MOD0 = MOD1 = matrix(0L, nrow = **ncell**, ncol = length(exp_genel))

I can get around this by putting an actual value for `select.ct` when calling the function, however, could you change the documentation to match the code, or code to match documentation?

**Comment by jaclynbeck-sage** [comment]
Also this same issue appears in `music2_prop_toast` and `music2_prop_t_statistics`.

**Comment by spatts14** [comment]
Hi @jaclynbeck-sage, what exactly you put for the select.ct value? Im assuming you just put all the cell types once, not label a list per cell - correct? Does it need to in the form of a list?

I got this error: Error in MOD0 + mod0 : non-conformable arrays
Calls: music2_prop_t_statistics

So Im trying to trouble shoot it . I had it set as NULL so it would use the single cell as default.

**Comment by jaclynbeck-sage** [comment]
@spatts14 Yes, I just put a vector of cell types: c('type1', 'type2', 'type3', ..., 'typeZ') (unique values, not a list per cell). That seemed to work ok.

**Comment by JoyOtten** [comment]
Hi I still have the same problem even when I put the cell types in a character vector.

**Comment by gevro** [comment]
Hi, I am getting the same issue.

**Comment by JoyOtten** [comment]
Hi, I think this issue has to do with the construction of your datasets. See here the code that I used for MuSiC (v1) version https://github.com/klengellab/Ska2/blob/main/Deconvolution_MuSic.R
This worked for me.

**Comment by gevro** [comment]
Thanks, but the issue is specific to music2 functions.  Music v1 works fine. So your code I don't think helps this issue.

**Comment by rob-meller** [comment]
Has there been any resolution to this issue.?  When I run my two data sets individually through MuSiC using a vector of cell_types, they work.  Then I run them together with MuSiC2 we get the "Error in MOD0 + mod0 : non-conformable arrays".  THX

**Comment by Porthmeus** [comment]
I had the same issue and I solved it as following. In the first step of MuSiC2 the proportions of cells in the control bulk samples are estimated (l. 77 in the code). In this step, the algorithm gave me a matrix of only 22 instead of the 23 cell type proportions I initially requested and this difference in dimensions causes the error described late in the code (l. 127). I simply removed the missing cell type from the `select.ct` argument and it worked. Interestingly, I also shortly tried to fix the issue in the source code - however, adding a column with the missing cell type in the matrix lead to other downstream issues, which I did not follow up on... Maybe one should reduce the vector of `select.ct` internally to fix it, which I have not explored. Anyway, I hope that this helps some people ending up here.

**Comment by WaterDrop-EarthDivision** [comment]
I had the same problem and after checking the previous replies I realized that it could be caused by a part of the cells having a low number in the single cell atlas, so I deleted the lesser number of epithelial cells in `select.ct` (using the table function to see that the number of cells of the species was only 2).

Hope this helps later 


```
est = music2_prop_t_statistics(bulk.control.mtx = bulk.control.mtx, bulk.case.mtx = bulk.case.mtx, sc.sce = sce, clusters = 'cell_type_level1', samples = 'sample',select. ct = c(‘Adipose’,'Immune',‘Stromal’,'Endothelial',‘Muscle’,'Neural',‘Epithelial ’), n_resample=20, sample_prop=0.5,cutoff_c=0.05,cutoff_r=0.01)



> table(sce$cell_type_level1)

    Adipose Endothelial Epithelial Immune Muscle Neural Stromal 
 31681 4281 2 846 50 112 1388



est = music2_prop_t_statistics(bulk.control.mtx = bulk.control.mtx, bulk.case.mtx = bulk.case.mtx, sc.sce = sce, clusters = ‘cell_type_level1’, samples = ‘sample’,select. ct = c(‘Adipose’,'Immune',‘Stromal’,'Endothelial',‘Muscle’,'Neural'), n_resample=20, sample_prop=0.5,cutoff_c= 0.05,cutoff_r=0.01)
```

**Comment by edwardgatesii3-ai** [comment]
> Running `music2_prop`, the following issue happens if 'select.ct' argument = NULL. It's allowed to be NULL in the original music_prop, and a NULL value means "use all cell types in the metadata" for that function. The documentation for `music2_prop` also says this. However there's no check or logic for the case where select.ct = NULL in `music2_prop`, so when it gets to the iteration loop it creates a MOD0 matrix with 0 rows:
> 
> ncell = length(select.ct) # returns 0 for select.ct = NULL ... MOD0 = MOD1 = matrix(0L, nrow = **ncell**, ncol = length(exp_genel))
> 
> I can get around this by putting an actual value for `select.ct` when calling the function, however, could you change the documentation to match the code, or code to match documentation?
